src/java/org/apache/cassandra/config/DataStorageSpec.java

```diff
@@ -38,117 +39,98 @@
      */
     private static final ImmutableMap<DataStorageUnit, ImmutableSet<DataStorageUnit>> MAP_UNITS_PER_MIN_UNIT =
-    ImmutableMap.of(KIBIBYTES, ImmutableSet.of(KIBIBYTES, MEBIBYTES, GIBIBYTES),
+    ImmutableMap.of(BYTES, ImmutableSet.of(BYTES, KIBIBYTES, MEBIBYTES, GIBIBYTES),
+                    KIBIBYTES, ImmutableSet.of(KIBIBYTES, MEBIBYTES, GIBIBYTES),
                     MEBIBYTES, ImmutableSet.of(MEBIBYTES, GIBIBYTES));
     /**
      * The Regexp used to parse the storage provided as String.
      */
-    private static final Pattern STORAGE_UNITS_PATTERN = Pattern.compile("^(\\d+)(GiB|MiB|KiB|B)$");
+    private static final Pattern UNITS_PATTERN = Pattern.compile("^(\\d+)(GiB|MiB|KiB|B)$");
 
     private final long quantity;
 
     private final DataStorageUnit unit;
 
-    public DataStorageSpec(String value)
+    private DataStorageSpec(long quantity, DataStorageUnit unit, DataStorageUnit smallestUnit, long max)
     {
-        //parse the string field value
-        Matcher matcher = STORAGE_UNITS_PATTERN.matcher(value);
-
-        if (!matcher.find())
-        {
-            throw new ConfigurationException("Invalid data storage: " + value + " Accepted units: B, KiB, MiB, GiB" +
-                                             " where case matters and only non-negative values are accepted");
-        }
-
-        quantity = Long.parseLong(matcher.group(1));
-        unit = DataStorageUnit.fromSymbol(matcher.group(2));
-    }
-
-    DataStorageSpec(long quantity, DataStorageUnit unit)
-    {
-        if (quantity < 0)
-            throw new ConfigurationException("Invalid data storage: value must be positive, but was " + quantity);
-
         this.quantity = quantity;
         this.unit = unit;
+
+        validateQuantity(quantity, unit, smallestUnit, max);
     }
 
-    public DataStorageSpec (String value, DataStorageUnit minUnit)
+    private DataStorageSpec (String value, DataStorageUnit minUnit)
     {
         if (!MAP_UNITS_PER_MIN_UNIT.containsKey(minUnit))
             throw new ConfigurationException("Invalid smallest unit set for " + value);
 
         //parse the string field value
-        Matcher matcher = STORAGE_UNITS_PATTERN.matcher(value);
+        Matcher matcher = UNITS_PATTERN.matcher(value);
 
         if (matcher.find())
         {
             quantity = Long.parseLong(matcher.group(1));
             unit = DataStorageUnit.fromSymbol(matcher.group(2));
 
+            //this constructor is used only by extended classes for smallest unit; upper bound is guarded there accordingly
+
             if (!MAP_UNITS_PER_MIN_UNIT.get(minUnit).contains(unit))
-                throw new ConfigurationException("Invalid data storage: " + value + " Accepted units:" + MAP_UNITS_PER_MIN_UNIT);
+                throw new ConfigurationException("Invalid data storage: " + value + " Accepted units:" + MAP_UNITS_PER_MIN_UNIT.get(minUnit));
```

Comment:
`DataStorageUnit` implements `Comparable`, so we could save us the `MAP_UNITS_PER_MIN_UNIT`  map of sets:
```java
if (unit.compareTo(minUnit) < 0)
    throw new ConfigurationException(String.format("Invalid data storage: %s Accepted units:%s",
                                                   value, acceptedUnits(minUnit)));
...
private String acceptedUnits(DataStorageUnit minUnit)
{
    DataStorageUnit[] units = DataStorageUnit.values();
    return Arrays.toString(Arrays.copyOfRange(units, minUnit.ordinal(), units.length));
}
```
That way we could add new units without manually updating that collection. Also, the simpler checks should probably be faster, albeit maybe slower on the uncommon case of building the exception messages.

Reply:
Not against that, all valid points, my reason was that with the set it is immediately clear while reading what we do instead of just comparing units. Probably your version is better. Let's see what Caleb thinks too and I can change it

Reply:
Also we wouldn't need the check at the beginning of the method:
```java
if (!MAP_UNITS_PER_MIN_UNIT.containsKey(minUnit))
            throw new ConfigurationException("Invalid smallest unit set for " + value);
```
I understand that the purpose of that check is verifying that the map of sets is well-formed and it contains every possible unit.

Reply:
It checks whether the unit provided is a valid min unit in general. But it seems it won't be needed if we switch to the version you suggested

Reply:
+1 looks like a nice simplification